server/src/main/java/org/opensearch/indices/replication/OngoingSegmentReplications.java

```diff
@@ -0,0 +1,195 @@
+             * the checkpoint from the created CopyState. This maximizes cache hits
+             * if replication targets make a request with an older checkpoint.
+             * Replication targets are expected to fetch the checkpoint in the response
+             * CopyState to bring themselves up to date.
+             */
+            addToCopyStateMap(checkpoint, copyState);
+            return copyState;
+        }
+    }
+
+    void cancelReplication(DiscoveryNode node) {
+        if (nodesToHandlers.containsKey(node)) {
+            final SegmentReplicationSourceHandler handler = nodesToHandlers.remove(node);
+            handler.cancel("Cancel on node left");
+            removeCopyState(handler.getCopyState());
+        }
+    }
+
+    SegmentReplicationSourceHandler createTargetHandler(DiscoveryNode node, CopyState copyState, FileChunkWriter fileChunkWriter) {
+        return new SegmentReplicationSourceHandler(
+            node,
+            fileChunkWriter,
+            copyState.getShard().getThreadPool(),
+            copyState,
+            Math.toIntExact(recoverySettings.getChunkSize().getBytes()),
+            recoverySettings.getMaxConcurrentFileChunks()
+        );
+    }
+
+    /**
+     * Adds the input {@link CopyState} object to {@link #copyStateMap}.
+     * The key is the CopyState's {@link ReplicationCheckpoint} object.
+     */
+    private void addToCopyStateMap(ReplicationCheckpoint checkpoint, CopyState copyState) {
+        copyStateMap.putIfAbsent(checkpoint, copyState);
+    }
+
+    /**
+     * Given a {@link ReplicationCheckpoint}, return the corresponding
+     * {@link CopyState} object, if any, from {@link #copyStateMap}.
+     */
+    private CopyState fetchFromCopyStateMap(ReplicationCheckpoint replicationCheckpoint) {
+        return copyStateMap.get(replicationCheckpoint);
+    }
+
+    /**
+     * Checks if the {@link #copyStateMap} has the input {@link ReplicationCheckpoint}
+     * as a key by invoking {@link Map#containsKey(Object)}.
+     */
+    boolean isInCopyStateMap(ReplicationCheckpoint replicationCheckpoint) {
+        return copyStateMap.containsKey(replicationCheckpoint);
+    }
+
+    void startSegmentCopy(GetSegmentFilesRequest request, ActionListener<GetSegmentFilesResponse> listener) {
+        final DiscoveryNode node = request.getTargetNode();
+        if (nodesToHandlers.containsKey(node)) {
+            final SegmentReplicationSourceHandler handler = nodesToHandlers.get(node);
+            // update the given listener to release the CopyState before it resolves.
+            final ActionListener<GetSegmentFilesResponse> wrappedListener = ActionListener.runBefore(listener, () -> {
+                final SegmentReplicationSourceHandler sourceHandler = nodesToHandlers.remove(node);
+                removeCopyState(sourceHandler.getCopyState());
+            });
+            handler.sendFiles(request, wrappedListener);
```

Comment:
You are right, the `handler` is not reusable, it is unique to a single replica to which we are sending files.  We reuse the `CopyState` between all shards, but the handler is unique.  This class manages both a cached map of `CopyState` objects and map of the handlers keyed by `DiscoveryNode`.  

The reason I'm storing these handlers in a map here is that we copy in two steps.  The replica first requests the metadata associated with a checkpoint and then computes a diff identifying the files it needs and makes another request to actually fetch the files.  The CopyState is created on the first request and stored so that we can ensure the files requested are not merged away between calls.  The handler is also created on the first request so that we can react to `clusterChanged` events when node is removed.  When that happens we want to decRef the copyState in addition to cancelling any ongoing transfer.

Reply:
Thanks @mch2 , the two-step processes was understood, perhaps we need additional guard here to know when the handler has initiated replication already and files are being replicated? Fe, when another `GetSegmentFilesRequest` comes in when there is one still in progress already for same node.

Reply:
Good call.  We've been handling duplicates on the replica side but better to have checks in both places.  I'm thinking in that case we noOp/error for any in progress replication.

Reply:
:+1: